### PR TITLE
Worked issue #10446 hide "obvious" inlays

### PR DESCRIPTION
**Pull request by emeryc**
The suggested method in the issue didn't work because that code path appears not to be hit during the generation of these types. Instead I introduced a new concept of unambigious types as part of let statements and filtered out cases that the issue seemed to be pointing at.

There were two tests that seemed to depend on the old behavior that I cleaned up in order to bring into alignment with the new behavior, this

This should solve #10446

### CODE REVIEW

File: crates/ide/src/inlay_hints.rs [hunk 1]
```diff
@@ -1154,7 +1206,6 @@ struct Test { a: Option<u32>, b: u8 }
 
 fn main() {
     let test = Some(Test { a: Some(3), b: 1 });
-      //^^^^ Option<Test>
```
Comment (Veykril) [inline crates/ide/src/inlay_hints.rs:1157]: These changes are undesirable, the idea behind this "hiding" is to only hide info if it is redundant. In this case here specifically, only `Some` is mentioned, so the fact that this local is of the type `Option<Test>` is not redundant and as such should always be shown.

**Review by Veykril — COMMENTED** [review]
> The suggested method in the issue didn't work because that code path appears not to be hit during the generation of these types. 

This sounds surprising to me, `is_named_constructor` should definitely be hit by for these cases. If they get filtered out earlier in this function call than that just has to be fixed so that they don't or such that they are properly handled earlier.

**Comment by emeryc** [comment]
call point is:
```rust
            if config.hide_named_constructor_hints
                && is_named_constructor(sema, pat, &ty_name).is_some()
```
hide_named_constructor_hints is false, and so we never evaluate the is_named_constructor due to short circuiting.

Should this be an ||

**Comment by Veykril** [comment]
Ah you need to enable the config for your test cases specifically as done here https://github.com/rust-analyzer/rust-analyzer/blob/9f447ad522a2e7464d827d5b51cdeb53a90c91f5/crates/ide/src/inlay_hints.rs#L1340-L1386

**Comment by Veykril** [comment]
@emeryc any updates here?

**Comment by emeryc** [comment]
Actual solution is going to be different enough for this PR to be not worth it. Sorry I should have just closed this sooner.

**Pull request closed without merging**